lib/src/deploy.rs

```diff
@@ -384,26 +386,50 @@ async fn deploy(
     } else {
         None
     };
-    // Because the C API expects a Vec<&str>, we need to generate a new Vec<>
-    // that borrows.
-    let override_kargs = override_kargs
-        .as_deref()
-        .map(|v| v.iter().map(|s| s.as_str()).collect::<Vec<_>>());
-    if let Some(kargs) = override_kargs.as_deref() {
-        opts.override_kernel_argv = Some(&kargs);
-    }
-    // Copy to move into thread
-    let cancellable = gio::Cancellable::NONE;
-    return sysroot
-        .stage_tree_with_options(
-            stateroot,
-            image.ostree_commit.as_str(),
-            Some(origin),
-            merge_deployment,
-            &opts,
-            cancellable,
-        )
-        .map_err(Into::into);
+    // Clone all the things to move to worker thread
+    let sysroot_clone = sysroot.sysroot.clone();
+    // ostree::Deployment is incorrently !Send 😢 so convert it to an integer
+    let merge_deployment = merge_deployment.map(|d| d.index() as usize);
+    let stateroot = stateroot.to_string();
+    let ostree_commit = image.ostree_commit.to_string();
+    // GKeyFile also isn't Send! So we serialize that as a string...
+    let origin_data = origin.to_data();
+    tracing::debug!("Deploying commit {ostree_commit}");
+    let r = async_task_with_spinner(
+        "Deploying",
+        spawn_blocking_cancellable_flatten(move |cancellable| -> Result<_> {
+            let sysroot = sysroot_clone;
+            let stateroot = Some(stateroot);
+            let mut opts = ostree::SysrootDeployTreeOpts::default();
+
+            // Because the C API expects a Vec<&str>, we need to generate a new Vec<>
+            // that borrows.
+            let override_kargs = override_kargs
+                .as_deref()
+                .map(|v| v.iter().map(|s| s.as_str()).collect::<Vec<_>>());
+            if let Some(kargs) = override_kargs.as_deref() {
+                opts.override_kernel_argv = Some(&kargs);
+            }
+            let deployments = sysroot.deployments();
+            let merge_deployment = merge_deployment.map(|m| &deployments[m]);
+            let origin = glib::KeyFile::new();
+            origin.load_from_data(&origin_data, glib::KeyFileFlags::NONE)?;
+            let d = sysroot.stage_tree_with_options(
+                stateroot.as_deref(),
+                &ostree_commit,
+                Some(&origin),
+                merge_deployment,
+                &opts,
+                Some(cancellable),
+            )?;
+            Ok(d.index())
+        }),
+    )
+    .await?;
+    // SAFETY: We must have a staged deployment
+    let staged = sysroot.staged_deployment().unwrap();
+    assert_eq!(staged.index(), r);
```

Comment:
Any reason to go for assertion/unwrap as opposed to `ensure!`/`?`

Reply:
I think the general guideline is:

- If the condition can ever be reached through external input, then use `?`
- If the condition represents an internal consistency check, `unwrap()` is OK

This is the latter: there's no way to change this externally. As the comments say we're just using an integer because the `Deployment` isn't `Send`, and we're verifying what we got back when re-parsing the state is what we expected. Nothing else could have changed that externally.

Reply:
Fair enough